compiler/rustc_borrowck/src/diagnostics/region_errors.rs

```diff
@@ -902,3 +945,46 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         suggest_adding_lifetime_params(self.infcx.tcx, sub, ty_sup, ty_sub, diag);
     }
 }
+
+/// Use of `Self` can impose unnecessary region constraints:
+/// ```
+/// struct MyStruct<'a>(&'a str);
+/// impl<'a> MyStruct<'a> {
+///     fn test<'x>(s: &'x str) {
+///         let _ = Self(s); // requires 'x: 'a.
+///     }
+/// }
+/// ```
+/// Here we suggest replacing `Self` with `MyStruct<'_>`.
+/// Assumes that `span` points to hir::Path that may name `Self`.
+pub(crate) fn suggest_relax_self<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    err: &mut Diagnostic,
+    body_did: DefId,
+    span: Span,
+) {
+    use rustc_lexer as lex;
+    fn tokenize(mut input: &str) -> impl Iterator<Item = (lex::TokenKind, &str)> {
+        std::iter::from_fn(move || {
+            if input.is_empty() {
+                return None;
+            }
+            let token = lex::first_token(input);
+            let token_str = &input[..(token.len as usize)];
+            input = &input[(token.len as usize)..];
+            Some((token.kind, token_str))
+        })
+    }
+
+    let snippet = tcx.sess.source_map().span_to_snippet(span).unwrap_or_default();
+    let has_self = tokenize(&snippet)
+        .find(|(tok, s)| *tok == lex::TokenKind::Ident && *s == "Self")
+        .and_then(|_| tcx.opt_parent(tcx.typeck_root_def_id(body_did)))
+        .filter(|parent_did| tcx.def_kind(parent_did) == DefKind::Impl);
```

Comment:
I was going to suggest try to get the hir id somehow, but that's not available in borrowck. *But* you can use the `body_did` and the span to find the right expr by creating a new `Visitor`:

https://github.com/rust-lang/rust/pull/101002/files#diff-23edf23f7466b97eb3154b82f7b324983287c327937fd66e854b3df6759926ceR645-R751